### ISSUE TITLE
Handle JSDoc `{@link ... }` inline tags

### ISSUE THREAD

**Issue by mjbvz** [microsoft/TypeScript#16498]
From: https://github.com/Microsoft/vscode/issues/28624

**TypeScript Version:**  2.3.5

**Code**

```ts
/**
 * @see {@link https://api.jquery.com/bind/#bind-eventType-eventData-handler}
 */
function foo() {
    
}
```

Hover over `foo`

**Bug**
The following docs are returned:

```
Result: {
    "kind": "function",
    "kindModifiers": "",
    "start": {
        "line": 4,
        "offset": 10
    },
    "end": {
        "line": 4,
        "offset": 13
    },
    "displayString": "function foo(): void",
    "documentation": "",
    "tags": [
        {
            "name": "see",
            "text": "{"
        },
        {
            "name": "link",
            "text": "https://api.jquery.com/bind/#bind-eventType-eventData-handler}"
        }
    ]
}
```

**Comment by kevinSuttle** [comment]
Just an FYI to VSCode users: you can use Markdown, which will render properly in the quick info. You won't get the syntax highlighting, though.

```
/**
 * [Link](https://example.com)
 */

**Comment by thw0rted** [comment]
Does this also address `{@link OtherClass.method}` ?  (Is that the right way to link to a member of another class that JSDoc knows about?)

**Comment by Duckers** [comment]
It would be awesome if links to language entities also worked with refactoring, so you can safely refactor entity without invalidating documentation that talks about those members.

**Comment by Rhysjc** [comment]
This doesn't seem to be fixed in the latest release. Any update?

**Comment by sandersn** [comment]
@Rhysjc we're still accepting PRs for this one.

This is primarily a parsing problem, and not an easy one, since `{@link ... }` doesn't look like any other JSDoc tags we support. It would be good if you're familiar with our parser, or are familiar with parsing in general and want to learn the quirks of ours.

Any solution will also have to decide whether `@link` is a child of whatever tag it appears in. The easy answer is 'no' since only a few tags can have children. But that would mean you would need to check that VS code displays a sequence of tags that includes `@link` in a pleasing way.

**Comment by DanielRosenwasser** [comment]
Linking to the VS UserVoice as well: https://visualstudio.uservoice.com/forums/121579-visual-studio-ide/suggestions/12796299-support-jsdoc-link-in-typescript

**Comment by couven92** [comment]
For what it is worth, I have written up my understanding of how the behaviour should be for rendering according to the cases specified on the Use JSDoc website. See my duplicate, now closed, issue: https://github.com/Microsoft/vscode/issues/57495#issue-355116663

**Comment by pvoisin** [comment]
What's going to happen with that issue? :)

**Comment by qkdreyer** [comment]
Same issue with PHPDoc :

![image](https://user-images.githubusercontent.com/717869/54539682-dd67cc00-4996-11e9-9596-a9376fbc2cf1.png)

**Comment by AnyhowStep** [comment]
Which files are involved with parsing JSDoc?

I doubt I'll make a dent in this but I am curious

**Comment by sandersn** [comment]
@AnyhowStep src/compiler/parser.ts. Start at parseJSDocComment, although `@link` is currently parsed inside `parseTagComments` because in #25206 we started treating `@link` as a comment.

**Comment by cancerberoSgx** [comment]
@thw0rted, I agree with @sandersn , like in other cases, existing TypeScript syntax should be used instead usejsdoc.org's. 

Would be awesome if I can use same syntax as when declaring a node's type or a type alias, like:

 * members: `@link {ClassA['member']}`
 * type helpers: `@link {ReturnType<typeof aFunction>}`
 * reference a parameter: `@link {Parameters<typeof f>[1]}` . This is interesting since authors normally reference parameters by name, and now will have to do it by index..

### Doubts 

 * should @link support expressions referencing multiple targets, like in : `@link {Foo<Bar>}`, ` @link {A | B | C}`, etc. I think it should. Implementations of the jsdoc renderer might have a hard time rendering this and probably forced to use the compiler API to find  definitions not just the AST ...

### Custom reference labeling

In general this feature also supports custom label / text so the same reference can be named with arbitrary text. **This project should not support that feature** and leave this responsibility to the description syntax language. For example, markdown descriptions could do it like `Lorem [Ipsum](@link {keyof Bar}) more text`, html descriptions like `Lorem <a href="@link {typeof foo}">custom label</a>  more text`, etc

### Relative references

many implementations support shorcuts to reference nodes by just their name and assume is the "closest" node. For example in a class description I can reference one of its members by just its name or in a function description I can reference a parameter with just its name. Supporting this is expensive and imply arbitrary decisions...

### Please please...

That will tackle formally/coherently most cases and the only missing part is to create a new node kind `JsDocLinkTag`that behaves just like JsDocReturnsTag. Authors will be forced to learn how TypeScript syntax or use type helpers but since the persona is a TypeScript developer that's fine

I think this feature is important and one of the few things missing not to support usejsdocs but for extensively documented APIs. And unfortunately there is no defactor standard for this, tsdoc, typedoc, jsdoc they all use different syntax. So this feature in this project will be very welcome, formalizing something very common which often rely on a informal syntax defined by the jsdoc renderer. Also IMO since TypeScript supports a big subset jsdocs at the syntax level (and there is no formal grammar for it), and since JavaScript IDEs use it to typechecking / intellicense JS code, its being now the defacto standard for jsdoc. So I think implementing this will be adopted quickly.

**Could a pull request implementing something like this accepted?** At least to start the discussion ? Or is something it won't be never implemented ? I assumed this "design" is the natural way for referencing nodes in descriptions, but maybe is not. 

I think this is not the same issue so should I open another one to describe this as a feature request?

Thanks.

**Comment by sandersn** [comment]
@cancerberoSgx You should create a separate issue for this since you have obviously put a lot of thought into it. Your previous comment is a good starting point since it already has tons of detail.

My first impression is that

1. We should not change `@link` since that's standard JSDoc.
2. The very first thing in your proposal should be the intended use case(s). I don't have a clear idea how this reference syntax will be used. I would like to see a description of a complete scenario, because right now I'm not convinced that Typescript is the best place to put this functionality.

**Comment by mjbvz** [comment]
Looks like TS now gives us back a better response:

```
[Trace  - 22:16:07.800] <semantic> Response received: quickinfo (171). Request took 1 ms. Success: true 
Result: {
    "kind": "function",
    "kindModifiers": "",
    "start": {
        "line": 4,
        "offset": 10
    },
    "end": {
        "line": 4,
        "offset": 13
    },
    "displayString": "function foo(): void",
    "documentation": "",
    "tags": [
        {
            "name": "see",
            "text": "{@link https://api.jquery.com/bind/#bind-eventType-eventData-handler}"
        }
    ]
}
```

@sandersn I'm going to try adding VS Code side logic to convert `{@link ...}` to real links. Do you think our TS should do anything to help editors parse these? One idea would be to return a set of link ranges in the response:

```
    "tags": [
        {
            "name": "see",
            "text": "{@link https://api.jquery.com/bind/#bind-eventType-eventData-handler|displayName}",
            "links": [{ start: 0, end: 20, text: 'displayName'}]
        }
```

**Comment by sandersn** [comment]
@mjbvz You're seeing the result of a change I made about a year ago to intentionally skip `@link`. `@link`, unlike every other JSDoc tag, is an inline tag, so I don't want to support it unless there are real, large benefits for TS users.

My opinion is that finding link spans would not remove a significant amount of work for TS API consumers, since those consumers are less likely to be constrained by performance requirements. Most consumers can probably afford a scan over a jsdoc JUST to find the `@link` spans, while the TS parser will have to track them as it parses.

@DanielRosenwasser What do you think?

**Comment by mjbvz** [comment]
I'm ok with trying to use a horrible regex to handle this on the VS Code side :) Just wasn't sure if TS already was parsing the inlines tags or not

/cc @amcasey @minestarks

**Comment by mjbvz** [comment]
VS Code 1.43 insiders will add support for `{@link }` with http/https uris. 

@sandersn We are still blocked on `{@link Symbol}` as editors cannot reasonably resolve the referenced symbol using the  current TSServer apis (one big  issue is that `Symbol` needs to be resolved in the scope of the definition, not at usage). We need TS to tell us where symbol links should go to

Related https://github.com/microsoft/TypeScript/issues/35524

**Comment by sandersn** [comment]
#35524 is a good start on a proposal. Let's track Typescript-based handling of `@link` there -- particularly binding.

**Comment by minestarks** [comment]
@mjbvz @sandersn just throwing it out there, if a "horrible regex to handle this" is on the table, could we at least put that handling in the language service (e.g. in the implementation of quickinfo, or wherever else we want to surface this) instead of VS Code? This way all editors get the benefit but we can still defer the tricker parsing questions to #35524.

**Comment by minestarks** [comment]
@mjbvz @sandersn  what is this issue supposed to be tracking at the moment? is it a dupe of #35524 ?

**Comment by sandersn** [comment]
Yes, I think it is at this point.

**Comment by rainmanhhh** [comment]
after 2 years {@link ...} still not work?

**Comment by brokenthorn** [comment]
> after 2 years {@link ...} still not work?

It's a complex issue. As you can see, it depends on TSServer functionality which is not ready. This is more complicated than it seems because of information that needs to come from context.

**Comment by babak1199** [comment]
Didn't see any mention of `{@link external:XMLHttpRequest}`. I hope it's part of the design too.

**Comment by leontepe** [comment]
The issue continues here: https://github.com/microsoft/TypeScript/issues/35524

**Comment by qJake** [comment]
The bottom of [this page](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#more-examples) should be updated with the new issue tracking link, if `{@link Symbol}` is still not yet supported.

People coming here from there (myself included), see this as closed and assume `{@link}` works across the board.

Is there a repo I can file a bug for that site to get it updated?

**Comment by minestarks** [comment]
@qjake https://github.com/microsoft/TypeScript-Website 
cc @orta

**Comment by orta** [comment]
Cool, I've updated the number on staging - it'll either get deployed some time this week or on the normal Monday deploy cycle.

**Comment by panlina** [comment]
Any update on this issue? This issue should not be closed.

**Comment by amcasey** [comment]
@Gaibhne @panlina Are you running typescript@next (aka nightly)?  As far as I know, this is not yet available in the the version shipped with VS Code.

**Comment by Gaibhne** [comment]
I'm using the regular stable release. As this issue was tagged as merged in May 2020, it would be merged for almost a whole year. Surely, it wouldn't still be stuck in nightly after all this time ?

**Comment by amcasey** [comment]
@Gaibhne I can't speak to the tagging, but the PR was only merged two weeks ago: https://github.com/microsoft/TypeScript/pull/43312

**Comment by MartinJohns** [comment]
This issue only related to `{@link ...}` tags containing http/https urls, and it works already. The other issue #35524 handles links to symbols and only got merged three weeks ago.

Just read the comments on this thread:
- https://github.com/microsoft/TypeScript/issues/16498#issuecomment-581703177
- https://github.com/microsoft/TypeScript/issues/16498#issuecomment-640237015

**Comment by MikeYermolayev** [comment]
> This issue only related to `{@link ...}` tags containing http/https urls, and it works already. The other issue #35524 handles links to symbols and only got merged three weeks ago.
> 
> Just read the comments on this thread:
> 
> * [Handle JSDoc `{@link ... }` inline tags #16498 (comment)](https://github.com/microsoft/TypeScript/issues/16498#issuecomment-581703177)
> * [Handle JSDoc `{@link ... }` inline tags #16498 (comment)](https://github.com/microsoft/TypeScript/issues/16498#issuecomment-640237015)

@MartinJohns looks like it doesn't work for me still. Links in hover popups aren't clickable and doesn't look like URLs
```
/**
 * React hook to support both controlled/uncontrolled modes
 * @see {@link https://reactjs.org/docs/uncontrolled-components.html}
 */
```

<img width="729" alt="image" src="https://user-images.githubusercontent.com/8783498/154353475-97a3619d-c0c5-45f3-972e-e7b2b5b5a8df.png">

**Comment by sandersn** [comment]
@MikeYermolayev please open a new issue for this. I believe the space after `https` is fixed in recent typescript builds, though, so check that out first.

**Comment by Gaibhne** [comment]
Both the original issue, the issue of it only working with brackets and the issue with the erroneous space are all fixed in the current version, so I am unsubscribing to this issue - thank you for fixing this!

**Comment by bradennapier** [comment]
@mjbvz 

There is currently a disconnect, i believe due to evolution of the `@link` definitions in the language service that creates a disconnect currently from `tsdoc` specification / `typedoc`

 I am not completely sure which side the disconnect is on, but the @link provides a LOT of opportunity to help users learn how to use a package as they actually code rather than referencing external website all the time.

I am leaning towards this being a `typedoc` bug (which is why the item below was opened there), but figured it may be worth looking into on the vscode side to utilize the language services features around the [Declaration References](https://typedoc.org/guides/declaration-references/)

> Screenshots of the disconnect in the issue linked below

- https://github.com/TypeStrong/typedoc/issues/2141

**Comment by sandersn** [comment]
@bradennapier for the vscode side of the example you give in the bug you link, name resolution is happening using normal typescript rules. I don't expect to change that. The typedoc resolution looks like a bug to me -- whether it's a spec bug or an implementation bug I'm not sure.

**Comment by bradennapier** [comment]
@sandersn yes that is as I thought, just wanted to make sure to mention here in case something seemed odd! 

Thanks for the reply, I appreciate your time.
